### PR TITLE
[ncp] add ncp spinel and implement GetDeviceRole

### PR DESCRIPTION
**Pull request by Irving-cl**
This PR implements NcpSpinel for the host to interact with the host.

In this PR, the first method `GetDeviceRole` is implemented and used
in NcpHost to get the device role.

### CODE REVIEW

File: src/ncp/ncp_spinel.cpp [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    case SPINEL_PROP_NET_ROLE:
+    {
+        spinel_net_role_t role;
+
+        unpacked = spinel_datatype_unpack(aBuffer, aLength, "C", &role);
+        VerifyOrExit(unpacked > 0, error = OTBR_ERROR_PARSE);
+
+        mDeviceRole = SpinelRoleToDeviceRole(role);
+
+        otbrLogInfo("Device role changed to %s", otThreadDeviceRoleToString(mDeviceRole));
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp:216]: `LogSpinelFrame` is already called in `SpinelDriver` (https://github.com/openthread/openthread/blob/main/src/lib/spinel/spinel_driver.cpp#L279), which means the these frames will be logged (no matter in NCP mode or RCP mode).

But LogSpinelFrame will only show when the log level is debug. Probably we can keep the log here or simply remove it. Thoughts?

File: src/ncp/ncp_spinel.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    VerifyOrExit(unpacked > 0, error = OT_ERROR_PARSE);
+
+    VerifyOrExit(cmd == SPINEL_CMD_PROP_VALUE_IS && key == mWaitingKeyTable[aTid], error = OT_ERROR_INVALID_STATE);
+
+    switch (key)
+    {
+    case SPINEL_PROP_NET_ROLE:
+    {
+        spinel_net_role_t spinelRole;
+        unpacked = spinel_datatype_unpack(data, len, "i", &spinelRole);
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp:178]: Updated.

File: src/ncp/ncp_spinel.cpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        break;
+    }
+
+exit:
+    FreeTid(aTid);
+    if (error == OT_ERROR_INVALID_STATE)
+    {
+        otbrLogCrit("Received unexpected response with cmd:%u, key:%u, waiting key:%u for tid:%u", cmd, key,
+                    mWaitingKeyTable[aTid], aTid);
+    }
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp:201]: Updated.

And in the cases where parsing fails (`VerifyOrExit` jumps to the `exit`), we still needs to call the handler with error code. So I introduced `FailureHandler`. In the handling of different keys, we set the `FailureHandler` first. And if the parsing isn't successful, the original handler will still be called in the end.

File: src/ncp/ncp_spinel.cpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    unpacked = spinel_datatype_unpack(aFrame, aLength, "CiiD", &header, &cmd, &key, &data, &len);
+    VerifyOrExit(unpacked > 0, error = OT_ERROR_PARSE);
+
+    VerifyOrExit(cmd == SPINEL_CMD_PROP_VALUE_IS && key == mWaitingKeyTable[aTid], error = OT_ERROR_INVALID_STATE);
+
+    switch (key)
+    {
+    case SPINEL_PROP_NET_ROLE:
+    {
+        spinel_net_role_t spinelRole;
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp:169]: Updated.

File: src/ncp/ncp_spinel.cpp [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+{
+    otError      error = OT_ERROR_NONE;
+    spinel_tid_t tid   = GetNextTid();
+    va_list      args;
+
+    error = mSpinelDriver->SendCommand(SPINEL_CMD_PROP_VALUE_GET, SPINEL_PROP_NET_ROLE, tid, nullptr, args);
+    if (error != OT_ERROR_NONE)
+    {
+        FreeTid(tid);
+        aHandler(error, OT_DEVICE_ROLE_DISABLED);
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp]: Thanks! Updated to use post to call the handler.

File: src/ncp/ncp_spinel.cpp [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    case SPINEL_PROP_NET_ROLE:
+    {
+        spinel_net_role_t spinelRole;
+        unpacked = spinel_datatype_unpack(data, len, "i", &spinelRole);
+
+        mDeviceRole = SpinelRoleToDeviceRole(spinelRole);
+
+        if (mGetDeviceRoleHandler)
+        {
+            mGetDeviceRoleHandler(error, mDeviceRole);
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp:185]: Updated.

File: src/ncp/ncp_spinel.cpp [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    : mSpinelDriver(nullptr)
+    , mCmdTidsInUse(0)
+    , mCmdNextTid(1)
+    , mDeviceRole(OT_DEVICE_ROLE_DISABLED)
+    , mGetDeviceRoleHandler(nullptr)
+{
+    memset(mWaitingKeyTable, 0xff, sizeof(mWaitingKeyTable));
+}
+
+NcpSpinel::~NcpSpinel(void) = default;
```
Comment (abtink) [inline src/ncp/ncp_spinel.cpp:55]: Better not including destructor definition if using `=default`. 
Can lead to poor code optimization by some tool chains (e.g. see [link](https://andreasfertig.blog/2023/11/why-you-shouldnt-provide-an-empty-destructor/))

File: src/ncp/ncp_spinel.cpp [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+void NcpSpinel::HandleResponse(spinel_tid_t aTid, const uint8_t *aFrame, uint16_t aLength)
+{
+    spinel_prop_key_t key;
+    spinel_size_t     len = 0;
+    spinel_ssize_t    unpacked;
+    uint8_t          *data = nullptr;
+    uint32_t          cmd;
+    uint8_t           header;
+    otError           error = OT_ERROR_NONE;
+    FailureHandler    failureHandler;
```
Comment (abtink) [inline src/ncp/ncp_spinel.cpp:158]: Does `FailureHandler` have a default constructor to set as `null` or empty?

File: src/ncp/ncp_spinel.cpp [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+
+    if (error == OT_ERROR_INVALID_STATE)
+    {
+        otbrLogCrit("Received unexpected response with cmd:%u, key:%u, waiting key:%u for tid:%u", cmd, key,
+                    mWaitingKeyTable[aTid], aTid);
+    }
+    else if (error == OT_ERROR_PARSE)
+    {
+        otbrLogCrit("Error parsing response with tid:%u", aTid);
+        failureHandler(error);
```
Comment (abtink) [inline src/ncp/ncp_spinel.cpp:194]: Should we ehck if `faiureHandler` is not `null`/empty? Can we call `std::function` it if it is empty?
Not sure if we can... https://en.cppreference.com/w/cpp/utility/functional/function/operator()

File: src/ncp/ncp_spinel.cpp (hunk 8 above)
Comment (abtink) [reply src/ncp/ncp_spinel.cpp:158]: Check`std::function<>`. Seems like it does init as empty.

**Review by abtink — APPROVED** [review]
LGTM. Thanks. 
Couple of questions/suggestions below

File: src/ncp/ncp_spinel.cpp (hunk 8 above)
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp]: Thanks! Though it inits as empty, I think we'd better to explicitly init it for better readability. Have added an explicit initialization.

File: src/ncp/ncp_spinel.cpp (hunk 9 above)
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp]: Yeah we should add a check. I've added a check here.

File: src/ncp/ncp_spinel.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/ncp/ncp_spinel.cpp:225]: Thanks for the explanation. I think we can keep the log here.

**Review by zhanglongxia — APPROVED** [review]
LGTM

File: src/ncp/ncp_spinel.cpp [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    error = mSpinelDriver->SendCommand(SPINEL_CMD_PROP_VALUE_GET, SPINEL_PROP_NET_ROLE, tid, nullptr, args);
+    if (error != OT_ERROR_NONE)
+    {
+        FreeTid(tid);
+        mTaskRunner.Post([aHandler, error](void) { aHandler(error, OT_DEVICE_ROLE_DISABLED); });
+        ExitNow();
+    }
+
+    mWaitingKeyTable[tid] = SPINEL_PROP_NET_ROLE;
+    mGetDeviceRoleHandler = aHandler;
```
Comment (superwhd) [inline src/ncp/ncp_spinel.cpp:81]: What's the expected behavior if there's already a `mGetDeviceRoleHandler`?

File: src/ncp/ncp_spinel.cpp [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+{
+    spinel_prop_key_t key;
+    spinel_size_t     len = 0;
+    spinel_ssize_t    unpacked;
+    uint8_t          *data = nullptr;
+    uint32_t          cmd;
+    uint8_t           header;
+    otbrError         error = OTBR_ERROR_NONE;
+
+    unpacked = spinel_datatype_unpack(aFrame, aLength, "CiiD", &header, &cmd, &key, &data, &len);
```
Comment (superwhd) [inline src/ncp/ncp_spinel.cpp:134]: Better to make constants for the string literals

File: src/ncp/ncp_spinel.cpp [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+
+        otbrLogInfo("NCP last status: %s", spinel_status_to_cstr(status));
+        break;
+    }
+    case SPINEL_PROP_NET_ROLE:
+    {
+        spinel_net_role_t role;
+
+        unpacked = spinel_datatype_unpack(aBuffer, aLength, "C", &role);
+        VerifyOrExit(unpacked > 0, error = OTBR_ERROR_PARSE);
```
Comment (superwhd) [inline src/ncp/ncp_spinel.cpp:221]: nit: may wrap the function `spinel_datatype_unpack` and let it return an OT error.

File: src/ncp/ncp_spinel.hpp [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+     *
+     */
+    void GetDeviceRole(GetDeviceRoleHandler aHandler);
+
+private:
+    using FailureHandler = std::function<void(otError)>;
+
+    static constexpr uint8_t kMaxTids = 16;
+
+    template <typename Function, typename... Args> void CallAndClear(Function &aFunc, Args &&...aArgs)
```
Comment (superwhd) [inline src/ncp/ncp_spinel.hpp:100]: It could be static

File: src/ncp/ncp_spinel.cpp (hunk 12 above)
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp]: Updated.

File: src/ncp/ncp_spinel.cpp (hunk 10 above)
Comment (Irving-cl) [reply src/ncp/ncp_spinel.cpp:83]: Considering the user of `NcpSpinel` is `NcpHost` which is also internal code, I prefer letting the new one override the existing one and there will be only one call to the handler. (The previous one will be discarded). I added some description in the API doc.

File: src/ncp/ncp_spinel.cpp [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+    VerifyOrExit(cmd == SPINEL_CMD_PROP_VALUE_IS && key == mWaitingKeyTable[aTid], error = OTBR_ERROR_INVALID_STATE);
+
+    switch (key)
+    {
+    case SPINEL_PROP_NET_ROLE:
+    {
+        spinel_net_role_t spinelRole;
+        failureHandler = [this](otError aError) { CallAndClear(mGetDeviceRoleHandler, aError, mDeviceRole); };
+
+        SuccessOrExit(error = SpinelDataUnpack(data, len, "i", &spinelRole));
```
Comment (superwhd) [inline src/ncp/ncp_spinel.cpp:181]: nit: May also make `"i"` a constant.

File: src/ncp/ncp_spinel.hpp [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+     * @param[in]  aHandler   A handler to return the role.
+     *
+     */
+    void GetDeviceRole(GetDeviceRoleHandler aHandler);
+
+private:
+    using FailureHandler = std::function<void(otError)>;
+
+    static constexpr uint8_t kMaxTids = 16;
+    static const char       *kSpinelDataUnpackFormat;
```
Comment (superwhd) [inline src/ncp/ncp_spinel.hpp:102]: Do you need this constant to be part of the class?

**Review by superwhd — APPROVED** [review]
LGTM overall 👍

File: src/ncp/ncp_spinel.hpp [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+     * Returns the Co-processor version string.
+     *
+     */
+    const char *GetCoprocessorVersion(void) { return mSpinelDriver->GetVersion(); }
+
+    /**
+     * This method gets the device role and return the role through the handler.
+     *
+     * If this method is called again before the handler is called, the previous handler will be
+     * overriden and there will be only one call to the latest handler.
```
Comment (jwhui) [inline src/ncp/ncp_spinel.hpp:91]:
```suggestion
     * overridden and there will be only one call to the latest handler.
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**